File: src-tauri/src/models/friends.rs
// Friends feature models
use serde::{Deserialize, Serialize};

/// Relationship type between users
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RelationshipType {
    Partner,
    Friend,
}

/// User presence status
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PresenceStatus {
    Online,
    Away,
    InGame,
    Offline,
}

impl Default for PresenceStatus {
    fn default() -> Self {
        Self::Offline
    }
}

/// Memory type for shared memories
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MemoryType {
    Photo,
    Video,
    Voice,
    Note,
    Countdown,
    Milestone,
}

/// User profile
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct User {
    pub id: String,
    pub friend_code: String,
    pub username: String,
    pub avatar_url: Option<String>,
    pub partner_id: Option<String>,
    pub created_at: u64,
}

impl User {
    pub fn new(id: String, username: String, friend_code: String) -> Self {
        Self {
            id,
            friend_code,
            username,
            avatar_url: None,
            partner_id: None,
            created_at: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_millis() as u64,
        }
    }
}

/// Local user data (stored on client)
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct LocalUserData {
    pub id: Option<String>,
    pub friend_code: Option<String>,
    pub username: Option<String>,
    pub server_url: String,
    pub auth_token: Option<String>,
    pub partner_id: Option<String>,
    pub last_sync: u64,
}

/// Connection state for server sync
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ConnectionState {
    Disconnected,
    Connecting,
    Connected,
    Error,
}

impl Default for ConnectionState {
    fn default() -> Self {
        Self::Disconnected
    }
}

/// Server sync result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FriendsSyncResult {
    pub success: bool,
    pub timestamp: u64,
    pub has_new_data: bool,
    pub new_messages_count: usize,
    pub new_pokes_count: usize,
    pub error: Option<String>,
}

/// Offline action to be synced when connection is restored
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OfflineAction {
    pub id: String,
    pub action_type: OfflineActionType,
    pub payload: serde_json::Value,
    pub created_at: u64,
}

/// Types of actions that can be queued offline
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum OfflineActionType {
    SendMessage,
    SendPoke,
    CreateMemory,
    CreateCalendarEvent,
    UpdateCalendarEvent,
    DeleteCalendarEvent,
    DeleteMemory,
    UpdatePresence,
}

/// Friend relationship
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Friend {
    pub id: String,
    pub user_id: String,
    pub friend_user_id: String,
    pub relationship_type: RelationshipType,
    pub nickname: Option<String>,
    pub created_at: u64,
}

/// Friend with user details (for display)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FriendWithDetails {
    pub friend: Friend,
    pub user: User,
    pub presence: Option<Presence>,
}

/// User presence (real-time status)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Presence {
    pub user_id: String,
    pub status: PresenceStatus,
    pub current_game: Option<String>,
    pub game_start_time: Option<u64>,
    pub mood_message: Option<String>,
    pub performance_stats: Option<PerformanceSnapshot>,
    pub last_updated: u64,
    pub last_seen: u64,
}

impl Presence {
    pub fn new(user_id: String) -> Self {
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;
        Self {
            user_id,
            status: PresenceStatus::Offline,
            current_game: None,
            game_start_time: None,
            mood_message: None,
            performance_stats: None,
            last_updated: now,
            last_seen: now,
        }
    }
}

/// Performance snapshot for sharing
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PerformanceSnapshot {
    pub cpu_usage: f32,
    pub gpu_usage: f32,
    pub fps: Option<f32>,
    pub memory_usage: f32,
}

/// Shared memory (photos, videos, notes)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Memory {
    pub id: String,
    pub user_id: String,
    pub partner_id: String,
    pub memory_type: MemoryType,
    pub content_url: Option<String>,
    pub content_text: Option<String>,
    pub caption: Option<String>,
    pub target_date: Option<u64>,
    pub created_at: u64,
}

impl Memory {
    pub fn new(
        user_id: String,
        partner_id: String,
        memory_type: MemoryType,
    ) -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            user_id,
            partner_id,
            memory_type,
            content_url: None,
            content_text: None,
            caption: None,
            target_date: None,
            created_at: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_millis() as u64,
        }
    }
}

/// Simple message between partners
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub id: String,
    pub sender_id: String,
    pub receiver_id: String,
    pub content: String,
    pub created_at: u64,
    pub read_at: Option<u64>,
}

impl Message {
    pub fn new(sender_id: String, receiver_id: String, content: String) -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            sender_id,
            receiver_id,
            content,
            created_at: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_millis() as u64,
            read_at: None,
        }
    }
}

/// Calendar event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CalendarEvent {
    pub id: String,
    pub user_id: String,
    pub partner_id: String,
    pub title: String,
    pub description: Option<String>,
    pub datetime: u64,
    pub timezone: String,
    pub reminder_minutes: Option<u32>,
    pub is_recurring: bool,
    pub recurrence_pattern: Option<String>,
    pub created_at: u64,
}

impl CalendarEvent {
    pub fn new(
        user_id: String,
        partner_id: String,
        title: String,
        datetime: u64,
        timezone: String,
    ) -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            user_id,
            partner_id,
            title,
            description: None,
            datetime,
            timezone,
            reminder_minutes: Some(30),
            is_recurring: false,
            recurrence_pattern: None,
            created_at: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_millis() as u64,
        }
    }
}

/// Poke/reaction notification
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Poke {
    pub id: String,
    pub sender_id: String,
    pub receiver_id: String,
    pub emoji: String,
    pub created_at: u64,
}

impl Poke {
    pub fn new(sender_id: String, receiver_id: String, emoji: String) -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            sender_id,
            receiver_id,
            emoji,
            created_at: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_millis() as u64,
        }
    }
}

/// Gacha pull notification for sharing
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GachaPullNotification {
    pub user_id: String,
    pub game: String,
    pub item_name: String,
    pub rarity: u8,
    pub pity: u32,
    pub timestamp: u64,
}

/// Shared gacha stats for comparison
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct SharedGachaStats {
    pub game: String,
    pub total_pulls: usize,
    pub five_star_count: usize,
    pub four_star_count: usize,
    pub average_pity: f64,
    pub current_pity: u32,
    pub luck_score: f64,
}

/// Wishlist item
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WishlistItem {
    pub id: String,
    pub user_id: String,
    pub game: String,
    pub item_name: String,
    pub item_type: String,
    pub priority: u8,
    pub created_at: u64,
}

/// Friend request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FriendRequest {
    pub id: String,
    pub from_user_id: String,
    pub to_user_id: String,
    pub relationship_type: RelationshipType,
    pub status: FriendRequestStatus,
    pub created_at: u64,
}

/// Friend request status
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FriendRequestStatus {
    Pending,
    Accepted,
    Rejected,
    Cancelled,
}

/// Update presence request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdatePresenceRequest {
    pub status: Option<PresenceStatus>,
    pub current_game: Option<String>,
    pub mood_message: Option<String>,
    pub performance_stats: Option<PerformanceSnapshot>,
}

/// Create memory request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateMemoryRequest {
    pub memory_type: MemoryType,
    pub content_text: Option<String>,
    pub caption: Option<String>,
    pub target_date: Option<u64>,
}

/// Create calendar event request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateCalendarEventRequest {
    pub title: String,
    pub description: Option<String>,
    pub datetime: u64,
    pub timezone: String,
    pub reminder_minutes: Option<u32>,
    pub is_recurring: bool,
    pub recurrence_pattern: Option<String>,
}

/// WebSocket message types
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum WebSocketMessage {
    PresenceUpdate(Presence),
    Poke(Poke),
    NewMessage(Message),
    GachaPull(GachaPullNotification),
    FriendRequest(FriendRequest),
    FriendRequestAccepted(Friend),
    MemoryAdded(Memory),
    CalendarEventAdded(CalendarEvent),
    Heartbeat,
    Error { message: String },
}

/// Server response for partner presence
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServerPresenceResponse {
    pub user_id: String,
    pub username: String,
    pub status: String,
    pub current_game: Option<String>,
    pub mood_message: Option<String>,
    pub performance_stats: Option<PerformanceSnapshot>,
    pub last_updated: u64,
}

/// Server sync poll response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncPollResponse {
    pub timestamp: u64,
    pub presence: Option<ServerPresenceResponse>,
    pub messages: Vec<Message>,
    pub pokes: Vec<ServerPoke>,
    pub memories: Vec<Memory>,
    pub calendar_events: Vec<CalendarEvent>,
    pub has_new_data: bool,
}

/// Server poke with sender username
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ServerPoke {
    pub id: String,
    pub sender_id: String,
    pub receiver_id: String,
    pub emoji: String,
    pub created_at: u64,
    pub sender_username: Option<String>,
}

/// Server registration response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegisterResponse {
    pub id: String,
    pub friend_code: String,
    pub username: String,
    pub auth_token: String,
    pub partner_id: Option<String>,
}

/// Server validation response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidateResponse {
    pub valid: bool,
    pub user: Option<ValidatedUser>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidatedUser {
    pub id: String,
    pub friend_code: String,
    pub username: String,
}

/// Link partner response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinkPartnerResponse {
    pub success: bool,
    pub partner: Option<ValidatedUser>,
}

/// Sync state response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyncStateResponse {
    pub timestamp: u64,
    pub has_partner: bool,
    pub partner: Option<ValidatedUser>,
    pub presence: Option<ServerPresenceResponse>,
    pub recent_messages: Vec<Message>,
    pub unread_count: usize,
    pub memories_count: usize,
    pub upcoming_events: Vec<CalendarEvent>,
}

/// Gacha stats for sharing with partner
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SharedGachaStatsPayload {
    pub game: String,
    pub total_pulls: usize,
    pub five_star_count: usize,
    pub four_star_count: usize,
    pub average_pity: f64,
    pub current_pity: u32,
}

/// Partner gacha stats response from server
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PartnerGachaStatsResponse {
    pub partner_id: String,
    pub partner_username: String,
    pub stats: Vec<PartnerGachaStats>,
}

/// Individual partner gacha stats
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PartnerGachaStats {
    pub user_id: String,
    pub username: String,
    pub game: String,
    pub total_pulls: usize,
    pub five_star_count: usize,
    pub four_star_count: usize,
    pub average_pity: f64,
    pub current_pity: u32,
    pub updated_at: u64,
}

File: src/types/friends.ts
// Friends feature types

export type RelationshipType = 'partner' | 'friend';

export type PresenceStatus = 'online' | 'away' | 'in_game' | 'offline';

export type MemoryType = 'photo' | 'video' | 'voice' | 'note' | 'countdown' | 'milestone';

export type FriendRequestStatus = 'pending' | 'accepted' | 'rejected' | 'cancelled';

// User profile
export interface User {
  id: string;
  friend_code: string;
  username: string;
  avatar_url: string | null;
  partner_id: string | null;
  created_at: number;
}

// Local user data (stored on client)
export interface LocalUserData {
  id: string | null;
  friend_code: string | null;
  username: string | null;
  server_url: string;
  auth_token: string | null;
}

// Friend relationship
export interface Friend {
  id: string;
  user_id: string;
  friend_user_id: string;
  relationship_type: RelationshipType;
  nickname: string | null;
  created_at: number;
}

// Friend with user details (for display)
export interface FriendWithDetails {
  friend: Friend;
  user: User;
  presence: Presence | null;
}

// Performance snapshot for sharing
export interface PerformanceSnapshot {
  cpu_usage: number;
  gpu_usage: number;
  fps: number | null;
  memory_usage: number;
}

// User presence (real-time status)
export interface Presence {
  user_id: string;
  status: PresenceStatus;
  current_game: string | null;
  game_start_time: number | null;
  mood_message: string | null;
  performance_stats: PerformanceSnapshot | null;
  last_updated: number;
  last_seen: number;
}

// Shared memory (photos, videos, notes)
export interface Memory {
  id: string;
  user_id: string;
  partner_id: string;
  memory_type: MemoryType;
  content_url: string | null;
  content_text: string | null;
  caption: string | null;
  target_date: number | null;
  created_at: number;
}

// Simple message between partners
export interface Message {
  id: string;
  sender_id: string;
  receiver_id: string;
  content: string;
  created_at: number;
  read_at: number | null;
}

// Calendar event
export interface CalendarEvent {
  id: string;
  user_id: string;
  partner_id: string;
  title: string;
  description: string | null;
  datetime: number;
  timezone: string;
  reminder_minutes: number | null;
  is_recurring: boolean;
  recurrence_pattern: string | null;
  created_at: number;
}

// Poke/reaction notification
export interface Poke {
  id: string;
  sender_id: string;
  receiver_id: string;
  emoji: string;
  created_at: number;
}

// Gacha pull notification for sharing
export interface GachaPullNotification {
  user_id: string;
  game: string;
  item_name: string;
  rarity: number;
  pity: number;
  timestamp: number;
}

// Shared gacha stats for comparison
export interface SharedGachaStats {
  game: string;
  total_pulls: number;
  five_star_count: number;
  four_star_count: number;
  average_pity: number;
  current_pity: number;
  luck_score: number;
}

// Wishlist item
export interface WishlistItem {
  id: string;
  user_id: string;
  game: string;
  item_name: string;
  item_type: string;
  priority: number;
  created_at: number;
}

// Friend request
export interface FriendRequest {
  id: string;
  from_user_id: string;
  to_user_id: string;
  relationship_type: RelationshipType;
  status: FriendRequestStatus;
  created_at: number;
}

// Update presence request
export interface UpdatePresenceRequest {
  status?: PresenceStatus;
  current_game?: string;
  mood_message?: string;
  performance_stats?: PerformanceSnapshot;
}

// Create memory request
export interface CreateMemoryRequest {
  memory_type: MemoryType;
  content_text?: string;
  caption?: string;
  target_date?: number;
}

// Create calendar event request
export interface CreateCalendarEventRequest {
  title: string;
  description?: string;
  datetime: number;
  timezone: string;
  reminder_minutes?: number;
  is_recurring: boolean;
  recurrence_pattern?: string;
}

// WebSocket message types
export type WebSocketMessage =
  | { type: 'PresenceUpdate'; payload: Presence }
  | { type: 'Poke'; payload: Poke }
  | { type: 'NewMessage'; payload: Message }
  | { type: 'GachaPull'; payload: GachaPullNotification }
  | { type: 'FriendRequest'; payload: FriendRequest }
  | { type: 'FriendRequestAccepted'; payload: Friend }
  | { type: 'MemoryAdded'; payload: Memory }
  | { type: 'CalendarEventAdded'; payload: CalendarEvent }
  | { type: 'Heartbeat' }
  | { type: 'Error'; payload: { message: string } };

// Helper functions

export function getPresenceStatusColor(status: PresenceStatus): string {
  switch (status) {
    case 'online':
      return 'bg-green-500';
    case 'away':
      return 'bg-yellow-500';
    case 'in_game':
      return 'bg-purple-500';
    case 'offline':
    default:
      return 'bg-gray-500';
  }
}

export function getPresenceStatusText(status: PresenceStatus): string {
  switch (status) {
    case 'online':
      return 'Online';
    case 'away':
      return 'Away';
    case 'in_game':
      return 'In Game';
    case 'offline':
    default:
      return 'Offline';
  }
}

export function formatLastSeen(timestamp: number): string {
  if (!timestamp) return 'Never';

  const now = Date.now();
  const diff = now - timestamp;

  const seconds = Math.floor(diff / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);

  if (seconds < 60) return 'Just now';
  if (minutes < 60) return `${minutes}m ago`;
  if (hours < 24) return `${hours}h ago`;
  if (days < 7) return `${days}d ago`;

  return new Date(timestamp).toLocaleDateString();
}

export function formatGameDuration(startTime: number | null): string {
  if (!startTime) return '';

  const now = Date.now();
  const diff = now - startTime;

  const minutes = Math.floor(diff / 60000);
  const hours = Math.floor(minutes / 60);
  const remainingMinutes = minutes % 60;

  if (hours > 0) {
    return `${hours}h ${remainingMinutes}m`;
  }
  return `${minutes}m`;
}

export function formatCountdown(targetDate: number): { days: number; hours: number; minutes: number; isPast: boolean } {
  const now = Date.now();
  const diff = targetDate - now;
  const isPast = diff < 0;
  const absDiff = Math.abs(diff);

  const days = Math.floor(absDiff / (24 * 60 * 60 * 1000));
  const hours = Math.floor((absDiff % (24 * 60 * 60 * 1000)) / (60 * 60 * 1000));
  const minutes = Math.floor((absDiff % (60 * 60 * 1000)) / (60 * 1000));

  return { days, hours, minutes, isPast };
}

export function formatCountdownText(targetDate: number): string {
  const { days, hours, minutes, isPast } = formatCountdown(targetDate);

  if (isPast) {
    if (days > 0) return `${days}d ago`;
    if (hours > 0) return `${hours}h ago`;
    return `${minutes}m ago`;
  }

  if (days > 0) return `${days}d ${hours}h`;
  if (hours > 0) return `${hours}h ${minutes}m`;
  return `${minutes}m`;
}

export function getMemoryTypeIcon(type: MemoryType): string {
  switch (type) {
    case 'photo':
      return 'Image';
    case 'video':
      return 'Video';
    case 'voice':
      return 'Mic';
    case 'note':
      return 'FileText';
    case 'countdown':
      return 'Clock';
    case 'milestone':
      return 'Award';
    default:
      return 'File';
  }
}

// Common poke emojis
export const POKE_EMOJIS = ['❤️', '😘', '🤗', '💕', '🥺', '💖', '✨', '🌟', '💫', '🎉'];

// Default server URL
export const DEFAULT_SERVER_URL = 'https://atlas-api.kaic5504.com';

// Connection state
export type ConnectionState = 'disconnected' | 'connecting' | 'connected' | 'error';

// Sync result from server
export interface FriendsSyncResult {
  success: boolean;
  timestamp: number;
  has_new_data: boolean;
  new_messages_count: number;
  new_pokes_count: number;
  error: string | null;
}

// Validation response
export interface ValidateResponse {
  valid: boolean;
  user: {
    id: string;
    friend_code: string;
    username: string;
  } | null;
}

// Server presence response
export interface ServerPresenceResponse {
  user_id: string;
  username: string;
  status: string;
  current_game: string | null;
  mood_message: string | null;
  performance_stats: PerformanceSnapshot | null;
  last_updated: number;
}

// Shared gacha stats payload (for uploading to server)
export interface SharedGachaStatsPayload {
  game: string;
  total_pulls: number;
  five_star_count: number;
  four_star_count: number;
  average_pity: number;
  current_pity: number;
}

// Partner gacha stats (from server)
export interface PartnerGachaStats {
  user_id: string;
  username: string;
  game: string;
  total_pulls: number;
  five_star_count: number;
  four_star_count: number;
  average_pity: number;
  current_pity: number;
  updated_at: number;
}

// Partner gacha stats response
export interface PartnerGachaStatsResponse {
  partner_id: string;
  partner_username: string;
  stats: PartnerGachaStats[];
}
